### PR TITLE
Enum refactoring

### PR DESCRIPTION
**Pull request by michalmuskala**
* Avoid do_ names in enum implementation
* Refactor enum list functions to body-recursive where sensible
  The reasoning is the same as in #3811 and 99e0d8e

### CODE REVIEW

**Comment by whatyouhide** [comment]
1. out of curiosity, have you measured the changes to tail recursion?
2. you renamed many functions to `*_list`, but then `protocol_into`: maybe `into_protocol` then?

**Comment by lexmag** [comment]
@michalmuskala did you, after [style guide tweet](https://twitter.com/michalmuskala/status/835106103243456515), decide to start spying on me? 🤔
_Just kidding._ I did extremely close changes yesterday https://github.com/lexmag/elixir/commit/70651563e79c5fa0b9d10b17537fd43a3a4a27f4.
Let's fix `h` and `t` usage like in that commit.

> maybe `into_protocol` then?

I think it's better to have common prefix. 👍

**Comment by whatyouhide** [comment]
@lexmag you mean `list_` and `protocol_` then>

**Comment by michalmuskala** [comment]
I initially wanted to go for `_list` instead of `_list`. But there were already some helper functions defined using the `_list` naming and I didn't want to create excessive diffs, so I decided to follow what I found.

**Comment by josevalim** [comment]
I prefer `_list` because it is easier to find. `all_list` is more straight-forward to find then if we have many many functions starting with `list_`.

**Comment by lexmag** [comment]
@whatyouhide by saying common prefix I meant prefix of helper function and public function, for example `all_list` helper for `all?` public function.

**Comment by whatyouhide** [comment]
okay, that sounds good :). I agree, let's have `_list` and `into_protocol` then :).

**Comment by whatyouhide** [comment]
Ping @michalmuskala? :)

**Comment by josevalim** [comment]
@whatyouhide maybe we should go ahead and merge it and do the changes ourselves as this PR can easily get stale.

**Comment by michalmuskala** [comment]
I tried benchmarking the change to see if it really gives the expected benefits and saw some surprising answers - for `zip`, as expected the body recursive is faster for small lists (teste for 1_000 elements), and the tail recursive is faster for larger lists (tested for 10_000_000 elements) - in both cases the difference is 10-15%. But for 100_000 elements, the body recursive approach is 3 times slower 😱 - I was able to reproduce this result consistently. That's why I stopped working on this in hopes I'll be able to figure out what's going on, but unfortunately I wasn't yet.

**Comment by josevalim** [comment]
@michalmuskala I have seen the same numbers in the past, except I never noticed a drastic slow down for 100_000 elements. I would expect some slow downs for certain elements for maps, due to internal resizing, but not for lists.

We should probably merge the PR only with the renaming and keep operations tail recursive. Especially because Enumerable.reduce is body recursive, so it would probably be best to keep the same properties.

**Comment by josevalim** [comment]
Ping.

**Comment by michalmuskala** [comment]
I updated the namings to use `head` and `tail` for the variables.

Additionally, there are two new commits:
* `Enum.fetch` performs only a single enumerable dispatch, instead of two, as was the common case previously
* `Enum.filter/2` and `Enum.reject/2` are optimised for lists with an inline function, instead of for, which is 30% to 10% faster.

As to the body-recursive vs tail-recursive. We already have some functions body recursive, most notably `Enum.map/2`. Here's my reasoning:
* For small collections the performance doean't really matter
* For small to medium size collections (up to thpusands of items) body recursive is faster. This is the most common case.
* For large collections the performance loss of using body-recursive vs tail recursive is 10-15% (which is not a horrible outcome). This isn't very common case to handle collections of millions of items. With such collections there are usually other issues, that lead to implement the algorithms manually, so the defaults don't matter that much.

**Comment by OvermindDL1** [comment]
> instead of for, which is 30% to 10% faster.

Why is `for` not generating the same style of private functions so it should have the same speed?  I'm fairly sure erlang's list/binary comprehensions do that so the speed should be the same?

**Comment by michalmuskala** [comment]
Elixir's for handles all enumerables, not only lists.

Additionally, there was a guard erroneously removed from `group_by` in https://github.com/elixir-lang/elixir/commit/99e44a181fb3440ef572be2f9592c94aa0a99ebc#diff-6881431a92cd4e3ea0de82bf2338f8eaL1032 - the guard was used for dispatch between current and deprecated function.

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -723,28 +723,30 @@ defmodule Enum do
   end
 
   def fetch(enumerable, index) when is_integer(index) and index < 0 do
-    case Enumerable.count(enumerable) do
-      {:error, _module} ->
+    module = Enumerable.impl_for!(enumerable)
+    case module.count(enumerable) do
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: This change is backwards incompatible. It is ok for a module to return ANOTHER module which has the proper implementation. This change also defeats the whole point of returning `{:error, __MODULE__}`, which is to avoid two protocol dispatches.

File: lib/elixir/lib/enum.ex [hunk 2]
```diff
@@ -723,28 +723,30 @@ defmodule Enum do
...
-        fetch_enumerable(enumerable, count + index, Enumerable)
+        fetch_enumerable(enumerable, count + index, module)
     end
   end
 
   def fetch(enumerable, index) when is_integer(index) do
-    case Enumerable.count(enumerable) do
-      {:error, module} ->
+    module = Enumerable.impl_for!(enumerable)
+    case module.count(enumerable) do
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: Same here, please revert.

**Comment by josevalim** [comment]
Thanks @michalmuskala, I have added two final comments.

File: lib/elixir/lib/enum.ex (hunk 1 above)
Comment (michalmuskala) [inline lib/elixir/lib/enum.ex]: First, the documentation of the enumerable protocol talks about returning `{:error, __MODULE__}` only. To me, this means returning anything else than the protocol module is not allowed.

Second, without this, we are doing double dispatch:
* when `Enumerable.count/1` does a dispatch and call returns `{:ok, count}`.
* in `fetch_enumerable/3` we pass `Enumerable`
* `fetch_enumerable/3` does another dispatch.

The only time a double-dispatch does not happen in this function is when count returns `{:error, __MODULE__}`.

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: @michalmuskala we do a double dispatch on the worst case scenario only. You do a double dispatch on all cases. If an implementation doesn't want to perform a double dispatch, then they can always implement `count`.

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: @michalmuskala please revert the changes and we will discuss how to proceed regarding count and member? separately.

Comment (michalmuskala) [inline lib/elixir/lib/enum.ex]: Sure, I will revert.

Could you explain where the second dispatch happens in my implementation, though? The first is during `impl_for!` call, but I can't find the second one.

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: You are right. There is no double protocol dispatch on this version of the code. Apologies. We should stick with the current version.

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: And by current version I mean your version.

**Comment by michalmuskala** [comment]
I reverted the fetch change, and will open a separate PR.

File: lib/elixir/lib/enum.ex [hunk 3]
```diff
@@ -806,7 +806,7 @@ defmodule Enum do
   """
   @spec filter(t, (element -> as_boolean(term))) :: list
   def filter(enumerable, fun) when is_list(enumerable) do
-    for item <- enumerable, fun.(item), do: item
+    filter_list(enumerable, fun)
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: I assume we can use `:lists.filter/2` here, instead of defining extra function?

Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Ignore me, we need to handle falsey values.

**Comment by lexmag** [comment]
Great job @michalmuskala. 💛
